Fix TrieMap#{filterInPlace,mapValuesInPlace}

NthPortal opened the pull request:
Fix the behaviour of `TrieMap#filterInPlace` and
`TrieMap#mapValuesInPlace` to respect atomic entry changes.

Built on top of #9726

Fixes scala/bug#12443

NthPortal commented:
currently an inefficient but simpler implementation. i'm working on a faster one, but... I think we should merge the reliable one first, and think about optimisation later

NthPortal commented:
oops I'm going to overengineer a test again 😬

johnynek commented on src/library/scala/collection/mutable/Map.scala:
```diff
@@ -171,17 +171,19 @@ trait MapOps[K, V, +CC[X, Y] <: MapOps[X, Y, CC, _], +C <: MapOps[K, V, CC, C]]
-        if (!p(k, v)) {
-          this -= k
+    if (!isEmpty) this match {
+      case tm: concurrent.TrieMap[_, _] => tm.asInstanceOf[concurrent.TrieMap[K, V]].filterInPlaceImpl(p)
```
Why this pattern vs overriding the method on TrieMap?

NthPortal replied:
the method returns `this.type`, which is the type of its class in bytecode. thus, the override has a different return type, and it is not binary compatible

johnynek commented on src/library/scala/collection/mutable/Map.scala:
```diff
@@ -197,8 +199,9 @@ trait MapOps[K, V, +CC[X, Y] <: MapOps[X, Y, CC, _], +C <: MapOps[K, V, CC, C]]
     */
   def mapValuesInPlace(f: (K, V) => V): this.type = {
-    if (nonEmpty) this match {
+    if (!isEmpty) this match {
```
Why `!isEmpty` instead of nonEmpty?

NthPortal replied:
basically, despite the fact that the latter really ought to be always trivially optimisable, `!isEmpty` sometimes has better performance then `nonEmpty`. I don't recall when/where this discovery happened or was mentioned; the best I can find in reference to it is #7898

lrytz commented on src/library/scala/collection/concurrent/TrieMap.scala:
```diff
@@ -1008,6 +1008,24 @@ final class TrieMap[K, V] private (r: AnyRef, rtupd: AtomicReferenceFieldUpdater
   }
 
+  private[collection] def filterInPlaceImpl(p: (K, V) => Boolean): this.type = {
+    val it = readOnlySnapshot().iterator
```
could this use `iterator` (which only calls `readOnlySnapshot()` if `nonReadOnly`)?

NthPortal replied:
if we're trying to call `filterInPlace` on a read-only snapshot, something has gone horribly, horribly wrong. (this is actually something we thought of when working on it, and arguably something we should have left a comment about, but it's moot anyway due to the following.)

regardless, the implementation needs to be moved up to the `concurrent.Map` trait (see https://github.com/scala/bug/issues/12443#issuecomment-898899755), so it's just going to call `iterator` without `readOnlySnapshot()` after we push that change anyway. we're figuring out how not to duplicate the test for the two implementations before pushing the change though

lrytz commented on src/library/scala/collection/concurrent/TrieMap.scala:
```diff
@@ -1008,6 +1008,24 @@ final class TrieMap[K, V] private (r: AnyRef, rtupd: AtomicReferenceFieldUpdater
+    val it = readOnlySnapshot().iterator
+    while (it.hasNext) {
+      val (k, v) = it.next()
+      if (p(k, v)) remove(k, v) // does unnecessary equality checking
```
what's the unnecessary check? using `==` instead of `eq` when comparing the `v`?

NthPortal replied:
correct. unfortunately, the contract of `remove` seems to require `==`. at least given that we implement it for `JConcurrentMapWrapper` using a Java method that uses `equals`

lrytz reviewed (commented):
LGTM (let's merge https://github.com/scala/scala/pull/9726 first though)

NthPortal committed:
Fix concurrent.Map#{filterInPlace,mapValuesInPlace}

Fix the behaviour of `concurrent.Map#filterInPlace` and
`concurrent.Map#mapValuesInPlace` to respect atomic entry changes.

NthPortal commented:
@lrytz I think this has changed enough to need a re-review

lrytz commented on test/junit/scala/collection/concurrent/ConcurrentMapTester.scala:
```diff
@@ -0,0 +1,41 @@
+import scala.concurrent.duration.TimeUnit
+
+class ConcurrentMapTester[K, V](map: Map[K, V]) {
+  private[this] val tasks = new AtomicReference[Seq[Runnable]](Queue.empty)
```
There's no concurrent access to `tasks`, is the `AtomicReference` future-proofing for tests that would do so?

NthPortal replied:
honestly, it's an artifact of a substantially different tester design. I'll try to simplify it

NthPortal replied:
now a `ConcurrentMapTester` can be used more than once (i.e. `runTasks` then check the results, `runTasks` again, etc.)

lrytz reviewed (approved):
LGTM, thank you!

Rebased and added MiMa filters.

NthPortal commented on src/library/scala/collection/concurrent/Map.scala:
```diff
@@ -131,4 +131,22 @@ trait Map[K, V] extends scala.collection.mutable.Map[K, V] {
+    val it = iterator
+    while (it.hasNext) {
+      val (k, v) = it.next()
+      if (p(k, v)) remove(k, v)
```
oops I inverted the condition by accident

```suggestion
      if (!p(k, v)) remove(k, v)
```

NthPortal replied:
fixed in #9737
